src/geometry/GeoTrackView.i.hh

```diff
@@ -74,54 +75,108 @@ GeoTrackView& GeoTrackView::operator=(const DetailedInitializer& init)
         pos_ = init.other.pos_;
     }
     // Set up the next state and initialize the direction
-    vgnext_.Clear();
-    next_step_ = celeritas::numeric_limits<real_type>::quiet_NaN();
-    dir_       = init.dir;
+    dir_ = init.dir;
+
+    this->find_next_step();
     return *this;
 }
 
 //---------------------------------------------------------------------------//
 //! Find the distance to the next geometric boundary.
-CELER_FUNCTION void GeoTrackView::find_next_step()
+CELER_FUNCTION void
+GeoTrackView::find_next_step(vecgeom::VPlacedVolume const* pplvol)
 {
-    const vecgeom::LogicalVolume* logical_vol
-        = this->volume().GetLogicalVolume();
-    CELER_ASSERT(logical_vol);
+    CELER_ASSERT(pplvol);
+
     const vecgeom::VNavigator* navigator
-        = this->volume().GetLogicalVolume()->GetNavigator();
+        = pplvol->GetLogicalVolume()->GetNavigator();
     CELER_ASSERT(navigator);
 
-    next_step_
-        = navigator->ComputeStepAndPropagatedState(detail::to_vector(pos_),
-                                                   detail::to_vector(dir_),
-                                                   vecgeom::kInfLength,
-                                                   vgstate_,
-                                                   vgnext_);
+    if (this->is_outside())
+    {
+        // handling points outside of world volume
+        const real_type large = vecgeom::kInfLength;
+        next_step_            = pplvol->DistanceToIn(
+            detail::to_vector(pos_), detail::to_vector(dir_), large);
+        vgnext_.Clear();
+        if (next_step_ < large)
+        {
+            vgnext_.Push(pplvol);
+        }
+    }
+    else
+    {
+        next_step_
+            = navigator->ComputeStepAndPropagatedState(detail::to_vector(pos_),
+                                                       detail::to_vector(dir_),
+                                                       vecgeom::kInfLength,
+                                                       vgstate_,
+                                                       vgnext_);
+    }
```

Comment:
Sorry I didn't catch this in my last view, but looking over it again -- I think the whole point of defining a `find_next_step(WhateverVolume*)` was to avoid this kind of switch statement, which should only ever be called during initialization rather than while tracking through the geometry. Can you refactor so that stepping through the volume doesn't encounter a conditional like this?   I think it might just mean moving the `is_outside()` case  inside the `operator=` .

Reply:
Done.  However, a similar treatment needs to be done when tracking leaves the world volume.
I have created another function `find_next_step_outside()` to separate the two cases, and the `if()s` have been moved to `operator=`, but also to `move_volume()`.

Reply:
I think it's OK to prevent meaningful tracking after *leaving* the geometry, as long as you've started inside it. So basically the only time we need to worry about being outside is during initialization. I'll take a look at the updates. Thanks!